WIP: Initial draft for va_win implementation for Win32 libva node

sivileri opened the pull request.

lindongw committed:
add Linux ftrace mode for va trace, trace data format will be offload to trace tool.

Signed-off-by: Lindong Wu <lindong.wu@intel.com>

RachelChengRJ committed:
Add HierarchicalFlag & hierarchical_level_plus1 for AV1e.

Signed-off-by: Cheng, Rachel <rachel.cheng@intel.com>

oleksandr-pavlyk committed:
Maint: Removed waffle-io badge from README to fix broken link

Since waffle.io has shut down in 2019, the broken badge and its link
will not be fixed at the destination.

Removing the badge is the only resolution.

Fixes #609

Signed-off-by: Oleksandr Pavlyk <oleksandr.pavlyk@intel.com>

XinfengZhang committed:
libva 2.16.0.pre1

Signed-off-by: Carl Zhang <carl.zhang@intel.com>

evelikov-work committed:
drm: add radeon drm + radeonsi mesa combo

The radeonsi driver can work with both radeon and amdgpu. Add the extra
combo into the mapping table.

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
android: remove convoluted open_device() helper

The helper does some fancy stat() and S_ISCHR() checking, neither of
which is necessary. Inline the open() call and drop the repeated error
reporting.

v2: preserve sys/stat.h include

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
x11: add basic DRI3 support

DRI3 has been a thing for 5+ years, sadly libva never learned how to
query the server for the fd.

Add basic support for that, where we fall-back to the DRM module's drm
module name to driver mappings.

v2: setup the autotools build
v3: explicitly handle all node types in va_isDRI3Connected()
v4: enhance fd ownership comment, flip node-type switch arms.
v5: drop fd ownership comment, close() as applicable
v6: null check after XGetXCBConnection()

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
android: open() with O_CLOEXEC

Otherwise we'll get fd leaks

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
android: drop va_fool references

The code was deprecated (read removed) a while ago, but some references
were forgotten.

Cc: Carl Zhang <carl.zhang@intel.com>
Fixes: aa47124 ("remove the va_fool support")
Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
drm: remove VA_DRM_IsRenderNodeFd() helper

Reuse the upstream drmGetNodeTypeFromFd(), which has been part of libdrm
version 2.4.59 (we require 2.4.60).

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
meson: set HAVE_VA_X11 when applicable

Without the define, we'll end without the Xorg auth whenever the DRM
display is used.

v2: move to va/meson.build, consistently use WITH_X11
v3: use the correct define

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
x11: simplify the mapping table

Currently we store the key_len and use it while comparing.. Just flip
the comparison and drop the key_len - keys are ~10 characters so there's
no point in it.

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
drm: simplify the mapping table

Currently we store the key_len and use it while comparing.. Just flip
the comparison and drop the key_len - keys are ~10 characters so there's
no point in it.

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>
Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov-work committed:
drm: fallback to drm driver name == va driver name

Currently we have a bunch of dummy entries, where both driver names
match. In addition, there is at least one more PR open to add a similar
mapping.

In the DRI2 mapping, we already fallback to the original (DRI2 driver)
name, so do the same here.

Add a convenience va_DRM_GetDrmDriverName() helper while we're here.

v2: Actually return the drm name

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

evelikov-work committed:
drm: remove no-longer needed X == X mappings

As with last commit, we don't need the mapping where both names match.

Signed-off-by: Emil Velikov <emil.velikov@collabora.com>

jbeich committed:
ci: upgrade FreeBSD to 13.1

FreeBSD doesn't support /latest and /quarterly package repos on EOL
versions. 13.0 reaches EOL on 2022-08-31, so avoid CI breakage.

dvrogozh committed:
ci: add clang-15 coverage and rearrange runners

Ubuntu 18.04 runner is deprecated and needs to be removed by 12/1/22
Ubuntu 22.04 runner is now officially supported and we can add it
clang-15 is a latest clang compiler we can also add

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

dvrogozh committed:
va/x11/nvctl: fix Wdeprecated-non-prototype on close_display

Fixes: #620

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

dvrogozh committed:
ci: strengthen ci with -Werror

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

jenatali commented on va/meson.build:
```diff
@@ -268,6 +286,53 @@ if WITH_WAYLAND
+  install_headers(libva_win32_headers, subdir : 'va')
+
+  libva_exports = 'libva.def'
+  libva_exports_path = '@0@/@1@'.format(meson.current_source_dir(), libva_exports)
```
This isn't needed I'm pretty sure, you should be able to just pass `'libva.def'` to `vs_module_defs`.

jenatali commented on va/va.c:
```diff
@@ -36,8 +36,31 @@
+#define dlsym GetProcAddress
+#define dlclose FreeLibrary
+#define dlopen(a, b) LoadLibrary(a)
+#define library_t HMODULE
```
`typedef HMODULE library_t` instead of `#define`

jenatali commented on va/va.c:
```diff
@@ -36,8 +36,31 @@
+#define strtok_r strtok_s
+#define dlsym GetProcAddress
+#define dlclose FreeLibrary
+#define dlopen(a, b) LoadLibrary(a)
```
I can't speak for what Intel's preference will be, but generally, I'd prefer to introduce an abstraction that doesn't exactly match *either* native platform API, like you did with `library_t`, rather than pretending that the platforms support an API set that belongs to a different platform.

Also, instead of using `#define`, prefer `inline` functions.

jenatali commented on va/va_compat.c:
```diff
@@ -40,4 +40,9 @@ vaCreateSurfaces_0_32_0(
-VA_CPP_HELPER_ALIAS(vaCreateSurfaces, 0, 32, 0);
+
+#if defined(_WIN32)
+#pragma comment(linker, "/alternatename:vaCreateSurfaces_0_32_0=vaCreateSurfaces@VA_API_0.32.0")
```
I don't think you need this? Especially because this needs name mangling for x86 and Arm64EC.

jenatali commented on va/va_trace.c:
```diff
@@ -44,16 +44,31 @@
+#include <windows.h>
+#include <io.h>
+#include <time.h>
+#define mutex_t HANDLE
```
You want `CRITICAL_SECTION`, not kernel mutexes.

jenatali commented on va/win32/va_win32.c:
```diff
@@ -0,0 +1,154 @@
+    VADisplayContextP pDisplayContext
+)
+{
+    return (pDisplayContext != NULL &&
```
Without looking at any other backends, this seems like it should have some kind of `magic` check.

sivileri replied:
I think the way it works is that `va_newDisplayContext` adds the `VA_DISPLAY_MAGIC` on creation. Then `CHECK_DISPLAY` and `vaDisplayIsValid` are called by the va runtime to check the magic value, delegating the other backend specific checks to `pDisplayContext->vaIsValid`

jenatali replied:
That makes sense.

jenatali commented on va/win32/va_win32.c:
```diff
@@ -0,0 +1,154 @@
+        // TODO: If adapter LUID is specified, query from
+        // some registry that maps drivers to adapter LUIDs
+        // or fallback to default driver otherwise
+        *num_candidates = 1;
```
You can take a look at https://github.com/KhronosGroup/OpenCL-ICD-Loader/blob/main/loader/windows/icd_windows_dxgk.c for an example of how to look up drivers from the registry. You probably want to cache this during display context creation.

jenatali commented on va/win32/va_win32.h:
```diff
@@ -0,0 +1,88 @@
+VAStatus vaPutSurface(
+    VADriverContextP ctx,
+    VASurfaceID surface,
+    void* draw, /* HWND window handle */
```
Is there a reason not specify this is an `HWND`?

jenatali commented on va/va.c:
```diff
@@ -49,13 +77,22 @@
+#if defined(_WIN32)
+#define last_err_string_size 512
+char last_err_string[last_err_string_size];
+static char* dlerror() {
```
Consider renaming this too.

sivileri committed:
ci: Add msvc/mingw Windows CI

Added windows.yml and EnterDevShell.ps1
for MSVC/mingw building with ninja/meson backend.

Signed-off-by: Sil Vilerino <sivileri@microsoft.com>

sivileri committed:
meson: Add with_win32 option, makes libdrm non-mandatory on Win

libdrm dependency needs to be non-mandatory by default on Windows builds

Signed-off-by: Sil Vilerino <sivileri@microsoft.com>

sivileri committed:
pkgconfig: Add Win32 package for when WITH_WIN32 is enabled

Signed-off-by: Sil Vilerino <sivileri@microsoft.com>

sivileri committed:
va: Adds Win32 Node and Windows build support

Added Win32 Node:
    vaGetDisplayWin32 receives LUID* to specify the desired
    adapter on Win32 platforms (NULL for autoselection is allowed)

    vaPutSurface receives HWND as Windowing system construct
    Surfaces can be imported/exported using either NT handles
    or ID3D12Resource object pointers

    A default VA driver name (vaon12) will be provided for some
    scenarios like for example when no specific LUID is specified
    in vaGetDisplayWin32

    When an adapter LUID is specified in vaGetDisplayWin32,
    the VA driver name will be dynamically loaded from a registry.
    If no entry is found in the registry for the specified adapter,
    the default VA driver will be used instead.

Signed-off-by: Sil Vilerino <sivileri@microsoft.com>

sivileri committed:
va: Adds compat_win32 abstraction for Windows build and prepares va common code for windows build

Implements dl*, pthread_mutex_*, gettimeofday and strtok_r on Windows platform

Signed-off-by: Sil Vilerino <sivileri@microsoft.com>

sivileri committed:
va_trace: Add missing pthread_mutex_destroy

pthread_mutex_destroy must be called on context_mutex and resource_mutex

Signed-off-by: Sil Vilerino <sivileri@microsoft.com>
Reviewed-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

sivileri committed:
va_trace: va_TraceSurfaceAttributes should check the VASurfaceAttribMemoryType

In the VASurfaceAttribExternalBufferDescriptor docs, it states that the memory type must be referred to
to know which type is passed with this attribute or use the default if not specified.
This change checks if a memory type attibute is passed that does not match the VASurfaceAttribExternalBuffers type.
Without this change, other memory types will cause va_TraceSurfaceAttributes to access garbage memory.

Signed-off-by: Sil Vilerino <sivileri@microsoft.com>

sivileri committed:
va_trace: Add Win32 memory types in va_TraceSurfaceAttributes

Add support for printing VA_SURFACE_ATTRIB_MEM_TYPE_NTHANDLE and VA_SURFACE_ATTRIB_MEM_TYPE_D3D12_RESOURCE

Signed-off-by: Sil Vilerino <sivileri@microsoft.com>

evelikov committed:
meson: use the wayland-scanner.pc to locate the prog

Currently we look for wayland-scanner in the prefix defined in the
wayland-client.pc. This is off, especially if we consider
cross-compilation.

Use the wayland-scanner.pc and it's wayland-scanner variable instead.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
autotools: use the wayland-scanner.pc to locate the prog

Currently we look for wayland-scanner in the prefix defined in the
wayland-client.pc. This is off, especially if we consider
cross-compilation.

Use the wayland-scanner.pc and it's wayland-scanner variable instead.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
meson: request native wayland-scanner

Otherwise we'll pull the incorrect binary and compilation will fail.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
meson: use wayland-scanner private-code

The "code" option was deprecated with wayland-scanner 1.15. The version
is available even in very old distros Debian Stretch and Ubuntu Bionic
(18.04), so it's fine to depend on it.

Here we replace the code option with private-code, since the code is
meant only for internal consumption.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
autotools: use wayland-scanner private-code

The "code" option was deprecated with wayland-scanner 1.15. The version
is available even in very old distros Debian Stretch and Ubuntu Bionic
(18.04), so it's fine to depend on it.

Here we replace the code option with private-code, since the code is
meant only for internal consumption.

With that done, we can drop the no longer needed export.c workaround.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
README: drop mailing list, IRC and Slack

All three are effectively dead. In particular:
 - the ML has disappeared with the update to HyperKitty
 - Freenode has imploded - channel has not moved to Libera/OFTC
 - Slack is a ghost town - last two maintainers are not even on it

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

XinfengZhang committed:
Correct slight coding style in several new commits

Signed-off-by: Carl Zhang <carl.zhang@intel.com>

oleksandr-pavlyk committed:
Update README.md

Removed all badge links as out of date

XinfengZhang committed:
update NEWS for 2.16.0

Signed-off-by: Carl Zhang <carl.zhang@intel.com>

XinfengZhang committed:
libva 2.17.0.pre1

Signed-off-by: Carl Zhang <carl.zhang@intel.com>

dvrogozh committed:
ci: add libx11-xcb-dev libxcb-dri3-dev to trigger X11 backend build

We forgot to add new packages required by recently merged DRI3
support into our ci install steps. As a result we were building
w/o X11 actually. Fixing this now an re-enabling X11 backend build.

See: #635
See: ef1df02 ("x11: add basic DRI3 support")
Reviewed-by: Emil Velikov <emil.l.velikov@gmail.com>
Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>
